File: python/heritage/application/delta_hedger.py
# ============================================================
# 文件：delta_hedger.py
# 职责：Delta 对冲器 — 订阅成交事件，监控组合 Delta 敞口，
#       当 |Δ| 超过阈值时提交市价对冲单，并直接记账对冲成交。
#
# 模式：策略模式（Strategy Pattern）+ 命令模式（Command Pattern）
#   - IDeltaHedgeStrategy 可替换对冲逻辑（Gamma 对冲、Vega 对冲等）
#   - OrderSubmittedEvent 将对冲意图封装为命令对象
#
# 注意：对冲成交直接绕过事件总线记入 PositionManager，
#       以避免触发新的 TradeExecutedEvent 导致无限递归。
#
# 事件订阅：TradeExecutedEvent（订阅后同步更新持仓与 Delta）
# 事件发布：OrderSubmittedEvent（当 |Δ| 超阈值时发出对冲单）
# ============================================================

from __future__ import annotations
import math

from events.event_bus import EventBus
from events.events import (
    TradeExecutedEvent, OrderSubmittedEvent,
    MarketDataEvent, Side, OrderType,
)
from domain.instrument import Option
from domain.pricing_engine import IPricingEngine
from domain.position_manager import PositionManager


class DeltaHedger:
    """基于 Delta 阈值的自动对冲器。"""

    def __init__(
        self,
        bus:              EventBus,
        position_manager: PositionManager,
        pricing_engine:   IPricingEngine,
        options:          list[Option],
        underlying_id:    str,
        delta_threshold:  float = 0.5,  # |Δ| 超过此值时触发对冲
    ) -> None:
        self._bus              = bus
        self._position_manager = position_manager
        self._pricing_engine   = pricing_engine
        self._options          = options
        self._underlying_id    = underlying_id
        self._delta_threshold  = delta_threshold
        self._last_price       = 150.0  # 最近一次标的价格（默认开盘估计值）

    def register_handlers(self) -> None:
        """向 EventBus 注册成交事件处理器。"""
        self._bus.subscribe(TradeExecutedEvent, self._on_trade_executed)

    def update_market_price(self, price: float) -> None:
        """更新标的资产最新价格（由行情事件驱动）。"""
        self._last_price = price

    def _on_trade_executed(self, event: TradeExecutedEvent) -> None:
        """成交事件处理器。

        流程：
          1. 更新持仓（PositionManager）
          2. 计算各合约 Delta
          3. 计算组合 Delta：Δ_portfolio = Σ(Δ_i × position_i)
          4. 若 |Δ| > 阈值：发出 OrderSubmittedEvent，直接记账对冲成交
        """
        # ── 步骤 1：更新持仓 ─────────────────────────────
        self._position_manager.on_trade_executed(event)

        # ── 步骤 2-3：计算组合 Delta ──────────────────────
        delta_map       = self._compute_delta_map(self._last_price)
        portfolio_delta = self._position_manager.compute_portfolio_delta(delta_map)

        print(
            f"[Delta对冲] 组合 Delta = {portfolio_delta:.3f}"
            f"  (阈值 ±{self._delta_threshold})"
        )

        if abs(portfolio_delta) <= self._delta_threshold:
            print("[Delta对冲] Delta 在阈值内，无需对冲")
            return

        # ── 步骤 4：触发对冲 ──────────────────────────────
        hedge_qty  = round(abs(portfolio_delta))
        hedge_side = Side.Sell if portfolio_delta > 0.0 else Side.Buy

        order = OrderSubmittedEvent(
            instrument_id = self._underlying_id,
            side          = hedge_side,
            quantity      = hedge_qty,
            order_type    = OrderType.Market,
        )

        side_str = "卖出" if hedge_side == Side.Sell else "买入"
        print(
            f"[Delta对冲] *** 触发对冲！发送市价单: "
            f"{side_str} {hedge_qty} 股 {self._underlying_id}"
            f"  (当前 Delta={portfolio_delta:.3f})"
        )

        self._bus.publish(order)

        # ── 对冲成交直接记账（绕过事件总线，防止递归）───────
        fill_side = Side.Sell if hedge_side == Side.Buy else Side.Buy
        hedge_fill = TradeExecutedEvent(
            instrument_id = self._underlying_id,
            side          = fill_side,
            price         = self._last_price,
            quantity      = hedge_qty,
            timestamp     = event.timestamp,
        )
        self._position_manager.on_trade_executed(hedge_fill)
        print("[Delta对冲] 对冲成交已直接记账（绕过事件总线，防止递归）")

    def _compute_delta_map(self, current_price: float) -> dict[str, float]:
        """计算所有合约的 Delta 映射。

        标的资产 Delta 恒为 1.0；
        各期权 Delta 由 pricing_engine 计算。
        """
        deltas: dict[str, float] = {self._underlying_id: 1.0}
        for opt in self._options:
            result             = self._pricing_engine.price(opt, current_price)
            deltas[opt.id]     = result.delta
        return deltas

File: src/modules/seller/DeltaHedger.cpp
#include "DeltaHedger.hpp"
#include <iostream>
#include <iomanip>
#include <cmath>    // std::abs, std::round

// 模式：策略模式 — DeltaHedger 是可替换的风险管理策略
// 模式：命令模式 — 通过 OrderSubmittedEvent 封装对冲订单意图
namespace omm::application {

DeltaHedger::DeltaHedger(
    std::shared_ptr<events::EventBus>            bus,
    std::shared_ptr<domain::PositionManager>     position_manager,
    std::shared_ptr<domain::IPricingEngine>      pricing_engine,
    std::vector<OptionEntry>                     options,
    const std::string&                           underlying_id,
    double delta_threshold,
    HedgeMode mode)
    : bus_(std::move(bus))
    , position_manager_(std::move(position_manager))
    , pricing_engine_(std::move(pricing_engine))
    , options_(std::move(options))
    , underlying_id_(underlying_id)
    , delta_threshold_(delta_threshold)
    , last_price_(150.0)
    , hedge_mode_(mode) {}

void DeltaHedger::register_handlers() {
    // FillEvent — record positions ONLY; do NOT compute delta or hedge here.
    // Deferring the rebalance to MarketDataEvent ensures that all legs of a
    // multi-leg strategy (e.g. call + put straddle) are fully recorded before
    // we compute portfolio delta.  Without this deferral, the call fill fires
    // a buy-hedge and the put fill (same bar, same price) fires a sell-hedge,
    // netting to $0 PnL regardless of subsequent spot moves.
    bus_->subscribe<events::FillEvent>(
        [this](const events::FillEvent& evt) { this->on_fill(evt); }
    );
    // MarketDataEvent — once-per-bar rebalance after all bar fills settle.
    bus_->subscribe<events::MarketDataEvent>(
        [this](const events::MarketDataEvent& evt) { this->on_market_data(evt); }
    );
}

void DeltaHedger::update_market_price(double price) {
    last_price_ = price;
}

// ── on_fill: position accounting only ───────────────────────
void DeltaHedger::on_fill(const events::FillEvent& event) {
    position_manager_->on_fill(event);

    // Hedge fills update the underlying position but must not trigger another
    // rebalance cycle.
    if (event.producer == "hedge_order" || event.instrument_id == underlying_id_) {
        return;
    }

    needs_rebalance_ = true;
}

// ── on_market_data: once-per-bar delta rebalance ─────────────
void DeltaHedger::on_market_data(const events::MarketDataEvent& event) {
    last_price_ = event.underlying_price;

    if (!needs_rebalance_) return;
    needs_rebalance_ = false;

    auto delta_map       = compute_delta_map(last_price_);
    double portfolio_delta = position_manager_->compute_portfolio_delta(delta_map);

    if (std::abs(portfolio_delta) <= delta_threshold_) return;

    int hedge_qty = static_cast<int>(std::round(std::abs(portfolio_delta)));
    events::Side hedge_side = (portfolio_delta > 0.0)
        ? events::Side::Sell
        : events::Side::Buy;

    std::cout << "[Delta对冲] *** 触发对冲！"
              << (hedge_side == events::Side::Buy ? " 买入 " : " 卖出 ")
              << hedge_qty << " 股 " << underlying_id_
              << "  (Delta=" << std::fixed << std::setprecision(2)
              << portfolio_delta << "  spot=" << last_price_ << ")\n";

    events::OrderSubmittedEvent order{
        underlying_id_,
        hedge_side,
        hedge_qty,
        events::OrderType::Market
    };
    order.reference_price = last_price_;
    order.producer = "hedge_order";
    order.timestamp = event.timestamp;
    bus_->publish(order);
}

std::unordered_map<std::string, double>
DeltaHedger::compute_delta_map(double current_price) const {
    std::unordered_map<std::string, double> deltas;

    // Underlying delta is always 1.0
    deltas[underlying_id_] = 1.0;

    // options_ is vector<pair<fill_id, option_ptr>>.
    // Key by fill_id so it matches PositionManager keys ("ATM_CALL" etc.)
    const bool use_market_state = (market_iv_ > 0.0 && market_T_ > 0.0);
    for (const auto& entry : options_) {
        const std::string&                     fill_id = entry.first;
        const std::shared_ptr<domain::Option>& option  = entry.second;
        domain::PriceResult result;
        if (use_market_state) {
            const bool is_call =
                (option->option_type() == domain::OptionType::Call);
            result = (hedge_mode_ == HedgeMode::ROUGH_DELTA)
                ? pricing_engine_->price_with_rough_delta(
                      current_price, option->strike(),
                      market_T_, market_iv_, is_call)
                : pricing_engine_->price_at_iv(
                      current_price, option->strike(),
                      market_T_, market_iv_, is_call);
        } else {
            result = pricing_engine_->price(*option, current_price);
        }
        deltas[fill_id] = result.delta;
    }

    return deltas;
}

} // namespace omm::application
